### DISCUSSION TITLE
Clarify Craft intialisation order for modules/plugins

### DISCUSSION THREAD

**Discussion by baileydoestech** [craftcms/cms#9671] [Q&A] [answered]
We are looking for clarifcation of the loading process used by Craft, specfically the order in which plugins, Twig extensions etc are intialised.

This is in response to https://github.com/nystudio107/craft-seomatic/issues/937 where Twig intialisation is called twice resulting in an internal server error. 

We believe the following is happening:

1.  Custom module loads before other plugins
2.  Module throws an exception (in our case this was intentional)
3.  The exception rendering code uses Twig to do its thing, so Twig is initialized
4.  Plugin loading continues, and another plugin tries to register its Twig extension
5.  Twig has already been initialized, so extensions can't be registered, and an internal server error is thrown

To remedy this we have wrapped the particular section of the module causing the exception in `EVENT_AFTER_LOAD_PLUGINS`:

```
public function init (): void
    {
        parent::init();
        self::$plugin = $this;

        $this->setComponents([
            'security'=> SecurityService::class,
        ]);

        Event::on(
            Plugins::class,
            Plugins::EVENT_AFTER_LOAD_PLUGINS,
            function () {
                if (Craft::$app->getRequest()->isCpRequest && !Craft::$app->getConfig()->general->devMode) {
                    $this->security->checkIpAddress();
                }
            }
        );
}
```

If this is correct and expected then I assume this process (whatever this currently is) changed in Craft 3.7 as the above issue certainly didn't manifest until recently.

**Comment by brandonkelly** [comment, marked as answer]
If your module is listed in the `bootstrap` array in `config/app.php`, thes you are correct, it will load before plugins.

Here’s the general order of request initialization:

1. Load Yii extensions
2. Load bootstrapped modules
3. Load plugins
4. Trigger `EVENT_INIT` event

Nothing changed about this in 3.7 that I can think of. It’s worked this way since 3.0.

**Reply by brandonkelly** [reply]
Craft 4.3.5 added a new `ApplicationTrait::onInit()` method, which simplifies things a bit (#12439).
